lib/api/contracts/model_contract.rb

```diff
@@ -48,19 +48,40 @@ def self.attribute(*attributes, &block)
         end
       end
 
+      def writable_attributes
+        collect_ancestor_attributes(:writable_attributes)
+      end
+
       validate :readonly_attributes_unchanged
       validate :run_attribute_validations
 
       private
 
       def readonly_attributes_unchanged
-        changed_attributes = model.changed - self.class.writable_attributes
+        changed_attributes = model.changed - writable_attributes
 
         errors.add :error_readonly, changed_attributes unless changed_attributes.empty?
       end
 
       def run_attribute_validations
-        self.class.attribute_validations.each { |validation| instance_exec(&validation) }
+        attribute_validations.each { |validation| instance_exec(&validation) }
+      end
+
+      def attribute_validations
+        collect_ancestor_attributes(:attribute_validations)
+      end
+
+      # All of the contract information across the entire inheritance hierarchy has to be
+      # considered in order to properly enforce them.
```

Comment:
Proposal:

````
# Traverse ancestor hierarchy to collect contract information.
# This allows to define attributes on a common base class of two or more contracts.
````

In the end I'd try to explain the use case, because on the base class the reader might not realize that subclasses may have subclasses of their own.